platform/linux-generic/include/odp/drv/mmio.h

```diff
@@ -0,0 +1,179 @@
+/* Copyright (c) 2017, Linaro Limited
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier:     BSD-3-Clause
+ */
+
+/**
+ * @file
+ *
+ * API to access memory-mapped I/O.
+ */
+
+#ifndef ODPDRV_PLAT_MMIO_H_
+#define ODPDRV_PLAT_MMIO_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <odp/api/byteorder.h>
+
+/** @ingroup odpdrv_mmio ODPDRV MMIO
+ *  @{
+ */
+
+/* for use with type checkers such as sparse */
+#ifdef __CHECKER__
+/** @internal MMIO attribute */
+#define __odpdrv_mmio	__attribute__((noderef, address_space(2)))
+#else
+/** @internal MMIO attribute */
+#define __odpdrv_mmio
+#endif
+
+#define odpdrv_io_mb() __asm__ __volatile__("" ::: "memory")
+#define odpdrv_io_rmb() __asm__ __volatile__("" ::: "memory")
+#define odpdrv_io_wmb() __asm__ __volatile__("" ::: "memory")
+
+static inline void
+odpdrv_mmio_u8le_write(uint8_t value, volatile void __odpdrv_mmio *addr)
+{
+	odpdrv_io_wmb();
+	*(__odp_force volatile uint8_t *)addr = value;
+}
+
+static inline void
+odpdrv_mmio_u8be_write(uint8_t value, volatile void __odpdrv_mmio *addr)
+{
+	odpdrv_mmio_u8le_write(value, addr);
+}
+
+static inline void
+odpdrv_mmio_u16le_write(uint16_t value, volatile void __odpdrv_mmio *addr)
+{
+	odpdrv_io_wmb();
+	*(__odp_force volatile odp_u16le_t *)addr = odp_cpu_to_le_16(value);
```

Comment:
Aren't odp_xxx_to_yyy_NN() expand to inline function which uses cc builtin ? I guess cc is smart enough to optimize out all wrapper code and use builtin (or unmodified value) directly. Is there really a reason to duplicate byteorder_inlines.h here ?

Side note on performance impact:

I wouldn't consider MMIO accesses performance critical (in a sense that you can optimize it and get performance boost). What is basically happening when you access MMIO is that CPU starts read/write transaction which ends up at the chipset (PCIe root complex in fact), the chipset sends MMIO transaction (64 bytes + overhead) to the peripheral and then awaits response (completion) from the peripheral. All in all it takes hundreds of CPU cycles to complete, and CPU can't do anything during this time. In short: it's way more expensive than even cache miss on host RAM, and register size (8-64 bits) has no impact on the total transaction cost.

So, any code which is accessing MMIO  too often is poorly designed, basically. In mediated device drivers we try to access it as little as possible, usually 1-2 times per TX/RX burst.

Reply:
ODP APIs can only expand inline when `--enable-abi-compat=no` is used, which means the application is targeting an embedded environment. Take a look at `odp_packet.c` which supports inline forms of APIs. The idea is that the external API may be inlined or not when called by applications, but the implementation itself always uses the inline form since it's self-contained for a specific target.

For these conversions, ODP itself is going to be compiled for a specific endianness so you can use the GCC `__builtin_bswapNN()` routines directly as needed. This avoids any function calls. For example:

```
/* Set up the conversion functions we need for this host's endianness */
#if ODP_BYTE_ORDER == ODP_LITTLE_ENDIAN
#define _odp_u16le(x) (x)
#define _odp_u16be(x) __builtin_bswap16(x)
#define _odp_u32le(x) (x)
#define _odp_u32be(x) __builtin_bswap32(x)
#define _odp_u64le(x) (x)
#define _odp_u64be(x) __builtin_bswap64(x)
#else
#define _odp_u16le(x) __builtin_bswap16(x)
#define _odp_u16be(x) (x)
#define _odp_u32le(x) __builtin_bswap32(x)
#define _odp_u32be(x) (x)
#define _odp_u64le(x) __builtin_bswap64(x)
#define _odp_u64be(x) (x)
#endif

static inline void
odpdrv_mmio_u16le_write(uint16_t valie, volatile void __odpdrv_mmio *addr)
{
        odpdrv_io_wmb();
        *(__odp_force volatile odp_u16le_t *)addr = _odp_u16le(value);
}

...etc.
```
Since these routines are on performance-critical paths, eliminating the calls to the external APIs is likely to be measurably beneficial.